File: 124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        ninf = -float('inf')
        ret = [ninf]
        def path(root):
            if not root:
                return ninf
            l= path(root.left)
            r = path(root.right)
            y = root.val
            if l>0:
                y+=l
            if r>0:
                y+=r
            ret[0]=max(ret[0], y)
            return max(root.val, max(root.val+l, root.val+r))
        
        path(root)
        return ret[0]
        

File: 124. Binary Tree Maximum Path Sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {
	ret := math.MinInt32

	max := func(a, b int) int {
		if a < b {
			return b
		}
		return a
	}

	var dfs func(node *TreeNode) int
	dfs = func(node *TreeNode) int {
		if node == nil {
			return math.MinInt32
		}

		left := max(node.Val, node.Val+dfs(node.Left))
		right := max(node.Val, node.Val+dfs(node.Right))

		ret = max(ret, left+right-node.Val)
		return max(left, right)
	}
	dfs(root)

	return ret
}